cvat/apps/engine/rq.py

```diff
@@ -0,0 +1,527 @@
+# Copyright (C) CVAT.ai Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+from typing import Any, Callable, Optional, Protocol, Union
+from uuid import UUID
+
+import attrs
+from django.conf import settings
+from django.db.models import Model
+from django.utils import timezone
+from django_rq.queues import DjangoRQ
+from rq.job import Dependency as RQDependency
+from rq.job import Job as RQJob
+from rq.registry import BaseRegistry as RQBaseRegistry
+
+from cvat.apps.engine.types import ExtendedRequest
+
+from .models import RequestAction, RequestSubresource, RequestTarget
+
+
+class RQJobMetaField:
+    class UserField:
+        ID = "id"
+        USERNAME = "username"
+        EMAIL = "email"
+
+    class RequestField:
+        UUID = "uuid"
+        TIMESTAMP = "timestamp"
+
+    # common fields
+    FORMATTED_EXCEPTION = "formatted_exception"
+    REQUEST = "request"
+    USER = "user"
+    PROJECT_ID = "project_id"
+    TASK_ID = "task_id"
+    JOB_ID = "job_id"
+    LAMBDA = "lambda"
+    ORG_ID = "org_id"
+    ORG_SLUG = "org_slug"
+    STATUS = "status"
+    PROGRESS = "progress"
+    TASK_PROGRESS = "task_progress"
+    # export specific fields
+    RESULT_URL = "result_url"
+    RESULT = "result"
+    FUNCTION_ID = "function_id"
+    EXCEPTION_TYPE = "exc_type"
+    EXCEPTION_ARGS = "exc_args"
+    TMP_FILE = "tmp_file"
+
+
+class WithMeta(Protocol):
+    meta: dict[str, Any]
+
+
+class _AbstractRQMetaAttribute:
+    def __init__(
+        self, key: str, *, optional: bool = False, validator: Callable | None = None
+    ) -> None:
+        assert validator is None or callable(validator), "validator must be callable"
+        self._key = key
+        self._validator = validator
+        self._optional = optional
+
+
+class _GettableRQMetaAttribute(_AbstractRQMetaAttribute):
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        if self._optional:
+            return instance.meta.get(self._key)
+
+        return instance.meta[self._key]
+
+
+class _SettableRQMetaAttribute(_AbstractRQMetaAttribute):
+    def validate(self, value):
+        if value is None and not self._optional:
+            raise ValueError(f"{self._key} is required")
+        if value is not None and self._validator and not self._validator(value):
+            raise ValueError("Wrong type")
```

Comment:
IMO, if the validator can only signal that the type is wrong, then you might as well replace the `validator` parameter with something like `expected_type`.

Reply:
I prefer to keep `validator`, changed exception message

Reply:
Okay, works for me.